### PR TITLE
Add Futhark implementation

### PR DESCRIPTION
**Pull request by athas**
I'm at SC22 and noticed that BabelStream is pretty popular, and I really like the idea of polyglot benchmark suites.  So, here's an implementation in [Futhark](https://futhark-lang.org).  I don't know if this is too obscure a language, but it's easy to invoke from C and C++ and so can use the same tooling as all the C++ implementations.

```
$ cmake -Bbuild -H. -DMODEL=futhark -DFUTHARK_BACKEND=opencl
...
$ cmake --build build
...
$ $ build/futhark-stream
BabelStream
Version: 4.0
Implementation: Futhark (OpencL)
Running kernels 100 times
Precision: double
Array size: 268.4 MB (=0.3 GB)
Total size: 805.3 MB (=0.8 GB)
Function    MBytes/sec  Min (sec)   Max         Average
Copy        1319634.622 0.00041     0.00141     0.00081
Mul         1338486.404 0.00040     0.00176     0.00081
Add         1355367.262 0.00059     0.00193     0.00098
Triad       1364525.941 0.00059     0.00196     0.00102
Dot         1315743.983 0.00041     0.00178     0.00079

### CODE REVIEW

**Comment by Munksgaard** [comment]
Perhaps the CI scripts should be updated as well?

**Comment by athas** [comment]
The more exotic implementations also aren't tested in CI, so I intentionally didn't add that to keep the invasiveness of this PR low.  But I can do so if desired, no problem.  (Incidentally, someone really ought to make a GA action for Futhark.)

**Comment by tom91136** [comment]
> The more exotic implementations also aren't tested in CI, so I intentionally didn't add that to keep the invasiveness of this PR low. But I can do so if desired, no problem. (Incidentally, someone really ought to make a GA action for Futhark.)

We do have CIs setup for the Rust, Julia, and Java variant so it would be good to have it for Futhark too.
Since it's using CMake, it would follow the CI steps like all other C++ implementations, see <https://github.com/UoB-HPC/BabelStream/blob/main/src/ci-test-compile.sh> and <https://github.com/UoB-HPC/BabelStream/blob/main/.github/workflows/main.yaml>.

**Comment by athas** [comment]
Alright, I'll add a CI step for Futhark.

**Comment by tom91136** [comment]
Thanks for the PR and sorry for the late reply, I'm taking a look now (running benchmarks, etc), one thing I've noticed is the lack of a device enumeration API in the Futhark runtime (but apparently you can set devices, or even be presented a dialog in the OpenCL case), this was a bit problematic as I'm testing machines with more than one OpenCL platform. As a workaround, we may have to implement the device enumeration by replicating the logic in the OpenCL and CUDA models.

**Comment by athas** [comment]
It wouldn't be difficult to add.  I can take a swing at it.

**Comment by tom91136** [comment]
@athas Do have to say, the generated C code for multicore CPU is more readable than a good portion C libraries out there!

**Comment by athas** [comment]
I can just copy the device enumeration code from the `cuda` and `ocl` implementations if you would prefer to have full functionality.

**Comment by athas** [comment]
I have added device selection now.

**Comment by tom91136** [comment]
I've got benchmark results for a few platforms.
For Nvidia A100, it's on-par with the native CUDA/OpenCL implementation:

BabelStream 4 0 A100 40G (array size=4 3GB)


Which is excellent.
For the multicore backend, I think the runtime is lacking NUMA awareness.
On a local Ryzen 5900X (1 NUMA domain) machine with dual channel DDR4 3400MT, I'm seeing comparable performance with OpenMP:

```shell
>./build/omp-stream --arraysize 536870912
BabelStream
Version: 4.0
Implementation: OpenMP
Running kernels 100 times
Precision: double
Array size: 4295.0 MB (=4.3 GB)
Total size: 12884.9 MB (=12.9 GB)
Function    MBytes/sec  Min (sec)   Max         Average
Copy        25610.437   0.33541     0.36010     0.34216
Mul         25402.003   0.33816     0.35810     0.34445
Add         29010.697   0.44414     0.47372     0.45097
Triad       28878.115   0.44618     0.47433     0.45268
Dot         44377.658   0.19356     0.21538     0.20067
>./build/futhark-stream --arraysize 536870912
BabelStream
Version: 4.0
Implementation: Futhark (parallel CPU)
Running kernels 100 times
Precision: double
Array size: 4295.0 MB (=4.3 GB)
Total size: 12884.9 MB (=12.9 GB)
Function    MBytes/sec  Min (sec)   Max         Average
Copy        26107.059   0.32903     0.35351     0.33425
Mul         25928.510   0.33129     0.35788     0.33712
Add         29328.709   0.43933     0.46747     0.44469
Triad       28930.892   0.44537     0.47550     0.45077
Dot         44656.475   0.19236     0.20958     0.19728

```

Performance is still good on a dual socket Intel Xeon Gold 6338 (2 NUMA domains) system compared to OpenMP:

BabelStream 4 0 Xeon Gold 6338 2 (Triad kernel, array size=4 3GB)

But for a dual socket AMD EPYC 7713 (8 NUMA domains), the performance is quite poor:

BabelStream 4 0 EPYC 7713 (array size=4 3GB)

I didn't do the scaling here due to time constraints.

**Comment by athas** [comment]
Yes, the multicore backend is completely NUMA-ignorant.  The GPU backends are much more mature.

File: src/futhark/FutharkStream.cpp [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+// Copyright (c) 2022 Troels Henriksen
+// University of Copenhagen
+//
+// For full license terms please see the LICENSE file distributed with this
+// source code
+
+#include <cstdlib>  // For aligned_alloc
+#include "FutharkStream.h"
+
+#ifndef ALIGNMENT
```
[the remark this answers is not in the document]
Comment (tom91136) [reply src/futhark/FutharkStream.cpp:13]: @athas

File: src/futhark/model.cmake [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Use
+#
+#   cmake -Bbuild -H. -DMODEL=futhark -DFUTHARK_BACKEND=foo
+#
+# to use the Futhark backend, where 'foo' must be one of 'multicore',
+# 'c', 'opencl', or 'cuda'.  Defaults to 'multicore'.  Expects that
+# the Futhark compiler is found somewhere on PATH.
+
+set(FUTHARK_BACKEND "multicore" CACHE STRING "Futhark compiler backend")
```
[the remark this answers is not in the document]
Comment (tom91136) [reply src/futhark/model.cmake:9]: @athas

**Review by tom91136 — CHANGES_REQUESTED** [review]

File: src/futhark/model.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (athas) [reply src/futhark/model.cmake]: Done.

**Comment by athas** [comment]
Apparently [it is impossible for me to do so](https://github.com/orgs/community/discussions/5634) for weird GitHub reasons.  I can see about resolving the conflicts myself.

**Comment by tom91136** [comment]
Thanks for the merge,  I've approved it now, let's wait for CI.

File: .github/workflows/main.yaml [hunk 3]
```diff
@@ -177,6 +177,56 @@ jobs:
...
+      - name: Test compile hip     @ CMake 3.24
+        if: ${{ ! cancelled() }}
+        run: ./src/ci-test-compile.sh ./build hip all ${{ env.CMAKE_3_24_BIN }}
+      - name: Test compile dpcpp   @ CMake 3.24
+        if: ${{ ! cancelled() }}
+        run: ./src/ci-test-compile.sh ./build dpcpp all ${{ env.CMAKE_3_24_BIN }}
+      - name: Test compile hipsycl @ CMake 3.24
+        if: ${{ ! cancelled() }}
+        run: ./src/ci-test-compile.sh ./build hipsycl all ${{ env.CMAKE_3_24_BIN }}
+
```
Comment (tom91136) [inline .github/workflows/main.yaml:229]: Can't comment beyond the diff, but for the `test-futhark` entry check the following:
 - Set `runs-on` to 22.04, 18.04 is EOL so it won't find a runner and hang indefinitely 
 - Use checkout@v4 to silence warnings

**Review by tom91136 — CHANGES_REQUESTED** [review]
Minor changes to get the CI going again

**Comment by athas** [comment]
CI fails due to an unrelated job running out of disk space.  I wonder why that is not an issue on the `develop` branch.

**Comment by tom91136** [comment]
No worries, it's probably because we ran out of cache due to how big the compilers are (we download and untar NVHPC in the setup). Thanks again and sorry about the slow turnaround.

**Pull request merged**